# SemEval Task 9 POLAR

### Subtask 3 - Manifestation Identification

### Hyper-parameter tune POC

By: Kevin Mcmahon, Caleb Kumar

Mount Google Drive to allow access to files stored in the user's Drive. This command will prompt the user for authorization.



In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


**Saving the Data Directory**



In [5]:
base_dir = '/content/drive/MyDrive/SemEval2026/'
data_dir = base_dir + 'data/subtask3/'

### Import statements

In [6]:
!pip install optuna
!pip install -U transformers

import pandas as pd

from sklearn.metrics import recall_score, precision_score, f1_score
import numpy as np
import random
import math

import torch

from sklearn.metrics import f1_score

from transformers import (
    AutoTokenizer,
    AutoConfig,
    AutoModelForSequenceClassification,
    Trainer,
    TrainingArguments,
    DataCollatorWithPadding,
    EarlyStoppingCallback,
    set_seed,
)
from torch.utils.data import Dataset
import wandb
from transformers import AutoConfig, AutoModelForSequenceClassification

import optuna
from optuna.samplers import TPESampler

import gc

### Importing / Formatting Data

In [7]:
base_dir = '/content/drive/MyDrive/SemEval2026/'
data_dir = base_dir + 'data/subtask3/'

from sklearn.model_selection import train_test_split

languages = ["eng","arb","deu"]

train_dfs = {}
dev_dfs = {}

for lang in languages:
    train_dfs[lang] = pd.read_csv(data_dir + f"train/{lang}.csv")
    train_dfs[lang]["language"] = lang
    dev_dfs[lang] = pd.read_csv(data_dir + f"dev/{lang}.csv")
    dev_dfs[lang]["language"] = lang

train_full = pd.concat([train_dfs[lang] for lang in languages], ignore_index=True)
dev_full = pd.concat([dev_dfs[lang] for lang in languages], ignore_index=True)

# 80/20 split for train/validation, preserving language distribution
train, validation = train_test_split(
    train_full,
    test_size=0.2,
    random_state=42,
    stratify=train_full["language"]
)

### Defining Dataset Object

In [8]:
# Dataset class (already multi-label friendly)
class PolarizationDataset(torch.utils.data.Dataset):
    def __init__(self, texts, labels, tokenizer, max_length=128):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_length = max_length

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        text = self.texts[idx]
        label = self.labels[idx]
        encoding = self.tokenizer(
            text,
            truncation=True,
            padding=False,
            max_length=self.max_length,
            return_tensors='pt'
        )

        item = {key: encoding[key].squeeze() for key in encoding.keys()}
        # multi-label → float labels
        item['labels'] = torch.tensor(label, dtype=torch.float)
        return item

### Defining Evaluation Metric

In [9]:
from sklearn.metrics import f1_score, accuracy_score

def compute_metrics_multilabel(p):
    # p.predictions is a numpy array of logits: (num_examples, num_labels)
    logits = torch.tensor(p.predictions)
    probs = torch.sigmoid(logits).numpy()

    # Try a slightly lower threshold to avoid "all zeros" early on.
    # You can tune this later; 0.3–0.4 is often more sensible for imbalanced multilabel.
    threshold = 0.3
    preds = (probs >= threshold).astype(int)

    labels = p.label_ids

    f1_macro = f1_score(labels, preds, average="macro", zero_division=0)
    # This is *subset* accuracy (exact match over all 6 labels for each example)
    accuracy = accuracy_score(labels, preds)

    return {
        "f1_macro": f1_macro,
        "accuracy": accuracy,
    }

In [10]:
MODEL_NAME = "cardiffnlp/twitter-xlm-roberta-base"

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)

MAX_LEN = 128

label_cols = [
    "vilification",
    "extreme_language",
    "stereotype",
    "invalidation",
    "lack_of_empathy",
    "dehumanization",
]

train_dataset = PolarizationDataset(
    train["text"].tolist(),
    train[label_cols].values.tolist(),
    tokenizer,
    max_length=MAX_LEN,
)

val_dataset = PolarizationDataset(
    validation["text"].tolist(),
    validation[label_cols].values.tolist(),
    tokenizer,
    max_length=MAX_LEN,
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/652 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

### Defining Callback Metrics

This will allow us to evalaute the performance of the experiment so that we can make sure that the model is learning.

In [11]:
from transformers import TrainerCallback

class EpochMetricsCallback(TrainerCallback):
    def __init__(self, trainer, train_dataset, val_dataset, trial=None):
        super().__init__()
        self.trainer = trainer
        self.train_dataset = train_dataset
        self.val_dataset = val_dataset
        self.trial = trial
        self.best_val_f1 = 0.0

    def on_epoch_end(self, args, state, control, **kwargs):
        epoch = state.epoch
        epoch_str = f"{epoch:.0f}" if epoch is not None and not math.isnan(epoch) else "?"

        print(f"\n===== Epoch {epoch_str} =====")

        # ---- Train metrics ----
        train_metrics = self.trainer.evaluate(eval_dataset=self.train_dataset)
        train_loss = train_metrics.get("eval_loss", float("nan"))
        train_f1 = train_metrics.get("eval_f1_macro", float("nan"))
        train_acc = train_metrics.get("eval_accuracy", float("nan"))
        print(
            "Train - "
            f"loss={train_loss:.4f}, "
            f"F1={train_f1:.4f}, "
            f"Acc={train_acc:.4f}"
        )

        # ---- Validation metrics ----
        val_metrics = self.trainer.evaluate(eval_dataset=self.val_dataset)
        val_loss = val_metrics.get("eval_loss", float("nan"))
        val_f1 = val_metrics.get("eval_f1_macro", 0.0)
        val_acc = val_metrics.get("eval_accuracy", float("nan"))
        print(
            "Val   - "
            f"loss={val_loss:.4f}, "
            f"F1={val_f1:.4f}, "
            f"Acc={val_acc:.4f}"
        )
        print("=========================\n")

        # Track best validation F1 for this trial
        if val_f1 > self.best_val_f1:
            self.best_val_f1 = val_f1

        # Report to Optuna + pruning
        if self.trial is not None:
            step = int(epoch) if epoch is not None and not math.isnan(epoch) else state.global_step
            self.trial.report(val_f1, step=step)
            if self.trial.should_prune():
                print(f"Pruning trial at epoch {epoch_str} with val F1={val_f1:.4f}")
                raise optuna.exceptions.TrialPruned()


### Optuna Objective

Optuna is a hyper-parameter tuning framework used to automate the process of hyper-parameter tuning while also doing it in a more efficient way using math rather than guess & check. I have previously used this in my final project for Deep Learning. The objective sets which parameters should be tuned, what the bounds for tuning the parameters are and what is the goal of the "sudy" - max Macro F1 in our case. Optuna will automatically suggest new parameters to test given the perfromance of the previously tested parameters. It will keep trying new "trials" for as long as you set unless it hits a time limit that you set... important when working in Google Colab.

In [12]:
import optuna
import gc
import torch
from transformers import TrainingArguments, Trainer
from transformers import DataCollatorWithPadding

def build_model(dropout: float):
    config = AutoConfig.from_pretrained(
        MODEL_NAME,
        num_labels=len(label_cols),
        problem_type="multi_label_classification",
        hidden_dropout_prob=dropout,
        attention_probs_dropout_prob=dropout,
    )
    model = AutoModelForSequenceClassification.from_pretrained(
        MODEL_NAME,
        config=config,
    )
    return model

GLOBAL_SEED = 42  # pick any int you like and stick with it


def seed_everything(seed: int):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)
    # HuggingFace helper (sets some internal RNGs)
    set_seed(seed)


# assumes:
# - GLOBAL_SEED is defined (e.g., GLOBAL_SEED = 42)
# - seed_everything(seed: int) is defined
# - build_model(dropout: float) is defined
# - compute_metrics_multilabel is defined
# - train_dataset, val_dataset, tokenizer, MODEL_NAME, label_cols are defined


def objective(trial: optuna.trial.Trial) -> float:
    trial_seed = GLOBAL_SEED + trial.number
    seed_everything(trial_seed)

    # Narrow LR around ~1e-5
    learning_rate = trial.suggest_float(
        "learning_rate",
        5e-6,    # lower
        2e-5,    # upper
        log=True,
    )

    # You’ve seen good behavior by 4–8 epochs
    num_train_epochs = trial.suggest_int("num_train_epochs", 4, 8)

    # If VRAM is fine, you can even fix this to 16
    per_device_train_batch_size = trial.suggest_categorical(
        "per_device_train_batch_size",
        [16],
    )

    # WD around 0.066
    weight_decay = trial.suggest_float("weight_decay", 0.03, 0.08)

    # Warmup around 0.05
    warmup_ratio = trial.suggest_float("warmup_ratio", 0.02, 0.08)

    # Dropout around 0.15
    dropout = trial.suggest_float("dropout", 0.12, 0.22)

    model = build_model(dropout)
    training_args = TrainingArguments(
        output_dir="./subtask3_tmp",
        num_train_epochs=num_train_epochs,
        learning_rate=learning_rate,
        per_device_train_batch_size=per_device_train_batch_size,
        per_device_eval_batch_size=per_device_train_batch_size,
        save_strategy="no",
        logging_steps=600,
        report_to="none",
        weight_decay=weight_decay,
        warmup_ratio=warmup_ratio,
        metric_for_best_model="f1_macro",
        load_best_model_at_end=False,
        fp16=torch.cuda.is_available(),
        seed=trial_seed,
        data_seed=trial_seed,
    )

    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=val_dataset,
        compute_metrics=compute_metrics_multilabel,
        data_collator=DataCollatorWithPadding(tokenizer),
    )

    epoch_cb = EpochMetricsCallback(
        trainer=trainer,
        train_dataset=train_dataset,
        val_dataset=val_dataset,
        trial=trial,
    )
    trainer.add_callback(epoch_cb)

    print("Trainer device:", trainer.args.device)

    try:
        trainer.train()
        eval_results = trainer.evaluate()
        print("Final eval metrics:", eval_results)
        f1 = max(epoch_cb.best_val_f1, eval_results["eval_f1_macro"])
    finally:
        del trainer, model
        gc.collect()
        if torch.cuda.is_available():
            torch.cuda.empty_cache()

    return f1

### Run Optuna "Study"

Above we defined the study, here we are actually going to run it.

In [13]:
study_name = f"study_{MODEL_NAME}"

sampler = TPESampler(seed=GLOBAL_SEED)

study = optuna.create_study(
    direction="maximize",
    study_name=study_name,
    sampler=sampler,
    pruner=optuna.pruners.MedianPruner(n_warmup_steps=1),
)

print(f"Starting study '{study_name}' with a 4-hour timeout...")
study.optimize(objective, n_trials=20, timeout=14400)

print("Best F1:", study.best_value)
print("Best params:", study.best_trial.params)


[I 2025-11-25 15:48:24,657] A new study created in memory with name: study_cardiffnlp/twitter-xlm-roberta-base


Starting study 'study_cardiffnlp/twitter-xlm-roberta-base' with a 4-hour timeout...


pytorch_model.bin:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.424273,0.459365,0.505411
600,0.501500,No Log,No Log,No Log
924,0.501500,0.403164,0.503959,0.469156
1200,0.412500,No Log,No Log,No Log
1386,0.412500,0.398373,0.519776,0.470779
1800,0.380300,No Log,No Log,No Log
1848,0.380300,0.404473,0.519869,0.499459
2310,0.380300,0.414012,0.520952,0.507035
2400,0.355600,No Log,No Log,No Log
2772,0.355600,0.411128,0.526779,0.502165



===== Epoch 1 =====
Train - loss=0.4201, F1=0.5042, Acc=0.4926
Val   - loss=0.4243, F1=0.4594, Acc=0.5054


===== Epoch 2 =====
Train - loss=0.3822, F1=0.5671, Acc=0.4764
Val   - loss=0.4032, F1=0.5040, Acc=0.4692


===== Epoch 3 =====
Train - loss=0.3557, F1=0.5996, Acc=0.4848
Val   - loss=0.3984, F1=0.5198, Acc=0.4708


===== Epoch 4 =====
Train - loss=0.3328, F1=0.6310, Acc=0.5370
Val   - loss=0.4045, F1=0.5199, Acc=0.4995


===== Epoch 5 =====
Train - loss=0.3169, F1=0.6528, Acc=0.5513
Val   - loss=0.4140, F1=0.5210, Acc=0.5070


===== Epoch 6 =====
Train - loss=0.3039, F1=0.6676, Acc=0.5548
Val   - loss=0.4111, F1=0.5268, Acc=0.5022


===== Epoch 7 =====
Train - loss=0.2967, F1=0.6750, Acc=0.5578
Val   - loss=0.4193, F1=0.5315, Acc=0.5022


===== Epoch 8 =====
Train - loss=0.2948, F1=0.6788, Acc=0.5647
Val   - loss=0.4254, F1=0.5324, Acc=0.5092



[I 2025-11-25 15:58:32,590] Trial 0 finished with value: 0.5324099539910884 and parameters: {'learning_rate': 8.403604888695184e-06, 'num_train_epochs': 8, 'per_device_train_batch_size': 16, 'weight_decay': 0.06659969709057026, 'warmup_ratio': 0.05591950905182219, 'dropout': 0.13560186404424365}. Best is trial 0 with value: 0.5324099539910884.


Final eval metrics: {'eval_loss': 0.42540886998176575, 'eval_f1_macro': 0.5324099539910884, 'eval_accuracy': 0.5091991341991342, 'eval_runtime': 2.5115, 'eval_samples_per_second': 735.807, 'eval_steps_per_second': 46.187, 'epoch': 8.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.424880,0.430981,0.442641
600,0.511500,No Log,No Log,No Log
924,0.511500,0.412842,0.494092,0.412879
1200,0.432300,No Log,No Log,No Log
1386,0.432300,0.406147,0.504701,0.462121
1800,0.415400,No Log,No Log,No Log
1848,0.415400,0.408257,0.506750,0.465368



===== Epoch 1 =====
Train - loss=0.4351, F1=0.4299, Acc=0.4061
Val   - loss=0.4249, F1=0.4310, Acc=0.4426


===== Epoch 2 =====
Train - loss=0.4134, F1=0.5263, Acc=0.3902
Val   - loss=0.4128, F1=0.4941, Acc=0.4129


===== Epoch 3 =====
Train - loss=0.3963, F1=0.5437, Acc=0.4514
Val   - loss=0.4061, F1=0.5047, Acc=0.4621


===== Epoch 4 =====
Train - loss=0.3926, F1=0.5558, Acc=0.4644
Val   - loss=0.4083, F1=0.5067, Acc=0.4654



[I 2025-11-25 16:03:36,741] Trial 1 finished with value: 0.5067497102081066 and parameters: {'learning_rate': 6.207090305742937e-06, 'num_train_epochs': 4, 'per_device_train_batch_size': 16, 'weight_decay': 0.07330880728874675, 'warmup_ratio': 0.056066900704592526, 'dropout': 0.19080725777960456}. Best is trial 0 with value: 0.5324099539910884.


Final eval metrics: {'eval_loss': 0.4082570970058441, 'eval_f1_macro': 0.5067497102081066, 'eval_accuracy': 0.4653679653679654, 'eval_runtime': 2.5089, 'eval_samples_per_second': 736.58, 'eval_steps_per_second': 46.236, 'epoch': 4.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.429209,0.482031,0.398810
600,0.509100,No Log,No Log,No Log
924,0.509100,0.407261,0.497438,0.473485
1200,0.424500,No Log,No Log,No Log
1386,0.424500,0.411073,0.486547,0.504329
1800,0.402600,No Log,No Log,No Log
1848,0.402600,0.410154,0.504454,0.490260
2310,0.402600,0.404149,0.515131,0.484848
2400,0.383500,No Log,No Log,No Log
2772,0.383500,0.412748,0.507148,0.500000



===== Epoch 1 =====
Train - loss=0.4368, F1=0.4928, Acc=0.3736
Val   - loss=0.4292, F1=0.4820, Acc=0.3988


===== Epoch 2 =====
Train - loss=0.3988, F1=0.5418, Acc=0.4614
Val   - loss=0.4073, F1=0.4974, Acc=0.4735


===== Epoch 3 =====
Train - loss=0.3850, F1=0.5566, Acc=0.5130
Val   - loss=0.4111, F1=0.4865, Acc=0.5043


===== Epoch 4 =====
Train - loss=0.3702, F1=0.5879, Acc=0.5031
Val   - loss=0.4102, F1=0.5045, Acc=0.4903


===== Epoch 5 =====
Train - loss=0.3584, F1=0.5976, Acc=0.4970
Val   - loss=0.4041, F1=0.5151, Acc=0.4848


===== Epoch 6 =====
Train - loss=0.3515, F1=0.6140, Acc=0.5202
Val   - loss=0.4127, F1=0.5071, Acc=0.5000


===== Epoch 7 =====
Train - loss=0.3469, F1=0.6156, Acc=0.5332
Val   - loss=0.4143, F1=0.5120, Acc=0.5092


===== Epoch 8 =====
Train - loss=0.3460, F1=0.6196, Acc=0.5356
Val   - loss=0.4164, F1=0.5141, Acc=0.5108



[I 2025-11-25 16:13:39,778] Trial 2 finished with value: 0.5151305576733921 and parameters: {'learning_rate': 5.144736127521127e-06, 'num_train_epochs': 8, 'per_device_train_batch_size': 16, 'weight_decay': 0.07162213204002109, 'warmup_ratio': 0.03274034664069657, 'dropout': 0.13818249672071006}. Best is trial 0 with value: 0.5324099539910884.


Final eval metrics: {'eval_loss': 0.4164103865623474, 'eval_f1_macro': 0.5140502741673941, 'eval_accuracy': 0.5108225108225108, 'eval_runtime': 2.4824, 'eval_samples_per_second': 744.427, 'eval_steps_per_second': 46.728, 'epoch': 8.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.420345,0.466306,0.481602
600,0.497900,No Log,No Log,No Log
924,0.497900,0.408485,0.472996,0.501082
1200,0.415100,No Log,No Log,No Log
1386,0.415100,0.405626,0.506837,0.496212
1800,0.399600,No Log,No Log,No Log
1848,0.399600,0.401832,0.520050,0.485390
2310,0.399600,0.408995,0.512246,0.500000



===== Epoch 1 =====
Train - loss=0.4269, F1=0.4796, Acc=0.4559
Val   - loss=0.4203, F1=0.4663, Acc=0.4816


===== Epoch 2 =====
Train - loss=0.3973, F1=0.5292, Acc=0.5004
Val   - loss=0.4085, F1=0.4730, Acc=0.5011


===== Epoch 3 =====
Train - loss=0.3778, F1=0.5748, Acc=0.5028
Val   - loss=0.4056, F1=0.5068, Acc=0.4962


===== Epoch 4 =====
Train - loss=0.3684, F1=0.5866, Acc=0.4927
Val   - loss=0.4018, F1=0.5201, Acc=0.4854


===== Epoch 5 =====
Train - loss=0.3660, F1=0.5928, Acc=0.5141
Val   - loss=0.4090, F1=0.5122, Acc=0.5000



[I 2025-11-25 16:19:58,565] Trial 3 finished with value: 0.5200500964744159 and parameters: {'learning_rate': 6.4474876947936455e-06, 'num_train_epochs': 5, 'per_device_train_batch_size': 16, 'weight_decay': 0.05623782158161189, 'warmup_ratio': 0.04591670111852694, 'dropout': 0.14912291401980418}. Best is trial 0 with value: 0.5324099539910884.


Final eval metrics: {'eval_loss': 0.40899479389190674, 'eval_f1_macro': 0.5122461568552409, 'eval_accuracy': 0.5, 'eval_runtime': 2.5015, 'eval_samples_per_second': 738.747, 'eval_steps_per_second': 46.372, 'epoch': 5.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.421403,0.427877,0.512446
600,0.484100,No Log,No Log,No Log
924,0.484100,0.404871,0.498824,0.503788
1200,0.411900,No Log,No Log,No Log
1386,0.411900,0.396970,0.522091,0.478355
1800,0.383400,No Log,No Log,No Log
1848,0.383400,0.404638,0.522813,0.500000



===== Epoch 1 =====
Train - loss=0.4160, F1=0.4662, Acc=0.5000
Val   - loss=0.4214, F1=0.4279, Acc=0.5124


===== Epoch 2 =====
Train - loss=0.3840, F1=0.5572, Acc=0.5041
Val   - loss=0.4049, F1=0.4988, Acc=0.5038


===== Epoch 3 =====
Train - loss=0.3617, F1=0.5923, Acc=0.4915
Val   - loss=0.3970, F1=0.5221, Acc=0.4784


===== Epoch 4 =====
Train - loss=0.3555, F1=0.6043, Acc=0.5150
Val   - loss=0.4046, F1=0.5228, Acc=0.5000



[I 2025-11-25 16:25:03,019] Trial 4 finished with value: 0.5228125903596381 and parameters: {'learning_rate': 1.1677292338861152e-05, 'num_train_epochs': 4, 'per_device_train_batch_size': 16, 'weight_decay': 0.04460723242676091, 'warmup_ratio': 0.0419817105976215, 'dropout': 0.1656069984217036}. Best is trial 0 with value: 0.5324099539910884.


Final eval metrics: {'eval_loss': 0.40463754534721375, 'eval_f1_macro': 0.5228125903596381, 'eval_accuracy': 0.5, 'eval_runtime': 2.5462, 'eval_samples_per_second': 725.8, 'eval_steps_per_second': 45.559, 'epoch': 4.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.416143,0.462147,0.518398
600,0.475100,No Log,No Log,No Log
924,0.475100,0.400458,0.503671,0.501082
1200,0.387400,No Log,No Log,No Log
1386,0.387400,0.405596,0.514312,0.507576
1800,0.352100,No Log,No Log,No Log
1848,0.352100,0.411494,0.518649,0.520563



===== Epoch 1 =====
Train - loss=0.4055, F1=0.5121, Acc=0.5114
Val   - loss=0.4161, F1=0.4621, Acc=0.5184


===== Epoch 2 =====
Train - loss=0.3584, F1=0.5861, Acc=0.5164
Val   - loss=0.4005, F1=0.5037, Acc=0.5011


===== Epoch 3 =====
Train - loss=0.3281, F1=0.6362, Acc=0.5441
Val   - loss=0.4056, F1=0.5143, Acc=0.5076


===== Epoch 4 =====
Train - loss=0.3178, F1=0.6496, Acc=0.5558


[I 2025-11-25 16:30:04,861] Trial 5 pruned. 


Val   - loss=0.4115, F1=0.5186, Acc=0.5206

Pruning trial at epoch 4 with val F1=0.5186


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.429842,0.426452,0.489177



===== Epoch 1 =====
Train - loss=0.4336, F1=0.4437, Acc=0.4674


[I 2025-11-25 16:31:22,135] Trial 6 pruned. 


Val   - loss=0.4298, F1=0.4265, Acc=0.4892

Pruning trial at epoch 1 with val F1=0.4265


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.409438,0.487000,0.491342
600,0.484200,No Log,No Log,No Log
924,0.484200,0.395309,0.516841,0.477814
1200,0.398600,No Log,No Log,No Log
1386,0.398600,0.396255,0.530706,0.485390
1800,0.362600,No Log,No Log,No Log
1848,0.362600,0.398206,0.537706,0.491883
2310,0.362600,0.410287,0.530590,0.511905



===== Epoch 1 =====
Train - loss=0.4055, F1=0.5169, Acc=0.4774
Val   - loss=0.4094, F1=0.4870, Acc=0.4913


===== Epoch 2 =====
Train - loss=0.3638, F1=0.5850, Acc=0.4888
Val   - loss=0.3953, F1=0.5168, Acc=0.4778


===== Epoch 3 =====
Train - loss=0.3350, F1=0.6279, Acc=0.5196
Val   - loss=0.3963, F1=0.5307, Acc=0.4854


===== Epoch 4 =====
Train - loss=0.3177, F1=0.6476, Acc=0.5313
Val   - loss=0.3982, F1=0.5377, Acc=0.4919


===== Epoch 5 =====
Train - loss=0.3131, F1=0.6560, Acc=0.5585
Val   - loss=0.4103, F1=0.5306, Acc=0.5119



[I 2025-11-25 16:37:40,697] Trial 7 finished with value: 0.5377062534189273 and parameters: {'learning_rate': 1.5334644233752532e-05, 'num_train_epochs': 5, 'per_device_train_batch_size': 16, 'weight_decay': 0.03488360570031919, 'warmup_ratio': 0.06105398159072942, 'dropout': 0.16401524937396011}. Best is trial 7 with value: 0.5377062534189273.


Final eval metrics: {'eval_loss': 0.4102874994277954, 'eval_f1_macro': 0.5305903551745219, 'eval_accuracy': 0.5119047619047619, 'eval_runtime': 2.528, 'eval_samples_per_second': 731.011, 'eval_steps_per_second': 45.886, 'epoch': 5.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.426522,0.484048,0.433983
600,0.519800,No Log,No Log,No Log
924,0.519800,0.409412,0.516871,0.446970
1200,0.427000,No Log,No Log,No Log
1386,0.427000,0.401240,0.523770,0.459416
1800,0.398200,No Log,No Log,No Log
1848,0.398200,0.406864,0.510904,0.515152
2310,0.398200,0.404497,0.518072,0.496212
2400,0.385500,No Log,No Log,No Log
2772,0.385500,0.407792,0.520171,0.504870



===== Epoch 1 =====
Train - loss=0.4392, F1=0.4881, Acc=0.4073
Val   - loss=0.4265, F1=0.4840, Acc=0.4340


===== Epoch 2 =====
Train - loss=0.4016, F1=0.5507, Acc=0.4410
Val   - loss=0.4094, F1=0.5169, Acc=0.4470


===== Epoch 3 =====
Train - loss=0.3811, F1=0.5676, Acc=0.4618
Val   - loss=0.4012, F1=0.5238, Acc=0.4594


===== Epoch 4 =====
Train - loss=0.3695, F1=0.5834, Acc=0.5195
Val   - loss=0.4069, F1=0.5109, Acc=0.5152


===== Epoch 5 =====
Train - loss=0.3605, F1=0.5967, Acc=0.5139
Val   - loss=0.4045, F1=0.5181, Acc=0.4962


===== Epoch 6 =====
Train - loss=0.3583, F1=0.5999, Acc=0.5199
Val   - loss=0.4078, F1=0.5202, Acc=0.5049



[I 2025-11-25 16:45:13,069] Trial 8 finished with value: 0.5237702127828637 and parameters: {'learning_rate': 5.921671927688252e-06, 'num_train_epochs': 6, 'per_device_train_batch_size': 16, 'weight_decay': 0.03171942605576092, 'warmup_ratio': 0.07455922412472692, 'dropout': 0.1458779981600017}. Best is trial 7 with value: 0.5377062534189273.


Final eval metrics: {'eval_loss': 0.40779218077659607, 'eval_f1_macro': 0.5201707858934354, 'eval_accuracy': 0.5048701298701299, 'eval_runtime': 2.4807, 'eval_samples_per_second': 744.959, 'eval_steps_per_second': 46.762, 'epoch': 6.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.406316,0.492381,0.430195
600,0.479100,No Log,No Log,No Log
924,0.479100,0.396373,0.526235,0.468615
1200,0.397800,No Log,No Log,No Log
1386,0.397800,0.409196,0.496861,0.525974
1800,0.364600,No Log,No Log,No Log
1848,0.364600,0.397075,0.530820,0.498918
2310,0.364600,0.412220,0.528161,0.509740



===== Epoch 1 =====
Train - loss=0.4084, F1=0.5119, Acc=0.4046
Val   - loss=0.4063, F1=0.4924, Acc=0.4302


===== Epoch 2 =====
Train - loss=0.3697, F1=0.5776, Acc=0.4716
Val   - loss=0.3964, F1=0.5262, Acc=0.4686


===== Epoch 3 =====
Train - loss=0.3471, F1=0.6058, Acc=0.5468
Val   - loss=0.4092, F1=0.4969, Acc=0.5260


===== Epoch 4 =====
Train - loss=0.3239, F1=0.6378, Acc=0.5342
Val   - loss=0.3971, F1=0.5308, Acc=0.4989


===== Epoch 5 =====
Train - loss=0.3181, F1=0.6450, Acc=0.5464
Val   - loss=0.4122, F1=0.5282, Acc=0.5097



[I 2025-11-25 16:51:29,664] Trial 9 finished with value: 0.5308199383588222 and parameters: {'learning_rate': 1.2527031373763456e-05, 'num_train_epochs': 5, 'per_device_train_batch_size': 16, 'weight_decay': 0.05600340105889054, 'warmup_ratio': 0.05280261676059678, 'dropout': 0.1384854455525527}. Best is trial 7 with value: 0.5377062534189273.


Final eval metrics: {'eval_loss': 0.4122198820114136, 'eval_f1_macro': 0.5281607011329177, 'eval_accuracy': 0.5097402597402597, 'eval_runtime': 2.4795, 'eval_samples_per_second': 745.301, 'eval_steps_per_second': 46.783, 'epoch': 5.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.409427,0.435054,0.458874



===== Epoch 1 =====
Train - loss=0.4061, F1=0.4642, Acc=0.4498
Val   - loss=0.4094, F1=0.4351, Acc=0.4589

Pruning trial at epoch 1 with val F1=0.4351


[I 2025-11-25 16:52:46,550] Trial 10 pruned. 
Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.431210,0.407502,0.527056



===== Epoch 1 =====
Train - loss=0.4339, F1=0.4177, Acc=0.5005


[I 2025-11-25 16:54:03,464] Trial 11 pruned. 


Val   - loss=0.4312, F1=0.4075, Acc=0.5271

Pruning trial at epoch 1 with val F1=0.4075


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.425588,0.428289,0.518398



===== Epoch 1 =====
Train - loss=0.4261, F1=0.4602, Acc=0.4997


[I 2025-11-25 16:55:20,520] Trial 12 pruned. 


Val   - loss=0.4256, F1=0.4283, Acc=0.5184

Pruning trial at epoch 1 with val F1=0.4283


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.407405,0.482224,0.428030
600,0.488900,No Log,No Log,No Log
924,0.488900,0.389109,0.511438,0.437771
1200,0.400200,No Log,No Log,No Log
1386,0.400200,0.405206,0.533298,0.455087
1800,0.354600,No Log,No Log,No Log
1848,0.354600,0.408405,0.541723,0.477814
2310,0.354600,0.424190,0.526979,0.514610
2400,0.315900,No Log,No Log,No Log
2772,0.315900,0.426854,0.539169,0.508117



===== Epoch 1 =====
Train - loss=0.4102, F1=0.4975, Acc=0.4051
Val   - loss=0.4074, F1=0.4822, Acc=0.4280


===== Epoch 2 =====
Train - loss=0.3641, F1=0.5740, Acc=0.4518
Val   - loss=0.3891, F1=0.5114, Acc=0.4378


===== Epoch 3 =====
Train - loss=0.3321, F1=0.6269, Acc=0.4900
Val   - loss=0.4052, F1=0.5333, Acc=0.4551


===== Epoch 4 =====
Train - loss=0.3001, F1=0.6644, Acc=0.5313
Val   - loss=0.4084, F1=0.5417, Acc=0.4778


===== Epoch 5 =====
Train - loss=0.2749, F1=0.6936, Acc=0.5824
Val   - loss=0.4242, F1=0.5270, Acc=0.5146


===== Epoch 6 =====
Train - loss=0.2630, F1=0.7070, Acc=0.5795
Val   - loss=0.4269, F1=0.5392, Acc=0.5081


===== Epoch 7 =====
Train - loss=0.2611, F1=0.7105, Acc=0.5888
Val   - loss=0.4441, F1=0.5287, Acc=0.5206



[I 2025-11-25 17:04:09,094] Trial 13 finished with value: 0.5417228608435262 and parameters: {'learning_rate': 1.8391979017484293e-05, 'num_train_epochs': 7, 'per_device_train_batch_size': 16, 'weight_decay': 0.07892880017961398, 'warmup_ratio': 0.06485857144825011, 'dropout': 0.15875942040639607}. Best is trial 13 with value: 0.5417228608435262.


Final eval metrics: {'eval_loss': 0.44406241178512573, 'eval_f1_macro': 0.5287320978290962, 'eval_accuracy': 0.5205627705627706, 'eval_runtime': 2.5503, 'eval_samples_per_second': 724.632, 'eval_steps_per_second': 45.486, 'epoch': 7.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.412772,0.426715,0.486472



===== Epoch 1 =====
Train - loss=0.4158, F1=0.4571, Acc=0.4720


[I 2025-11-25 17:05:26,053] Trial 14 pruned. 


Val   - loss=0.4128, F1=0.4267, Acc=0.4865

Pruning trial at epoch 1 with val F1=0.4267


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.415865,0.492910,0.436688
600,0.494200,No Log,No Log,No Log
924,0.494200,0.409810,0.520846,0.465368
1200,0.401300,No Log,No Log,No Log
1386,0.401300,0.397219,0.527461,0.503247
1800,0.360800,No Log,No Log,No Log
1848,0.360800,0.409585,0.530463,0.515693
2310,0.360800,0.409350,0.541646,0.500541
2400,0.328400,No Log,No Log,No Log
2772,0.328400,0.435970,0.512473,0.526515



===== Epoch 1 =====
Train - loss=0.4125, F1=0.5245, Acc=0.4249
Val   - loss=0.4159, F1=0.4929, Acc=0.4367


===== Epoch 2 =====
Train - loss=0.3721, F1=0.5810, Acc=0.4690
Val   - loss=0.4098, F1=0.5208, Acc=0.4654


===== Epoch 3 =====
Train - loss=0.3308, F1=0.6306, Acc=0.5300
Val   - loss=0.3972, F1=0.5275, Acc=0.5032


===== Epoch 4 =====
Train - loss=0.3059, F1=0.6624, Acc=0.5558
Val   - loss=0.4096, F1=0.5305, Acc=0.5157


===== Epoch 5 =====
Train - loss=0.2882, F1=0.6809, Acc=0.5525
Val   - loss=0.4093, F1=0.5416, Acc=0.5005


===== Epoch 6 =====
Train - loss=0.2842, F1=0.6900, Acc=0.5887
Val   - loss=0.4360, F1=0.5125, Acc=0.5265


===== Epoch 7 =====
Train - loss=0.2734, F1=0.7001, Acc=0.5804
Val   - loss=0.4273, F1=0.5295, Acc=0.5097



[I 2025-11-25 17:14:11,746] Trial 15 finished with value: 0.5416464318523418 and parameters: {'learning_rate': 1.4840911587616934e-05, 'num_train_epochs': 7, 'per_device_train_batch_size': 16, 'weight_decay': 0.04713996441331581, 'warmup_ratio': 0.07075312433244194, 'dropout': 0.15763868440634574}. Best is trial 13 with value: 0.5417228608435262.


Final eval metrics: {'eval_loss': 0.42730772495269775, 'eval_f1_macro': 0.5295252156884478, 'eval_accuracy': 0.5097402597402597, 'eval_runtime': 2.5113, 'eval_samples_per_second': 735.882, 'eval_steps_per_second': 46.192, 'epoch': 7.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.407320,0.501240,0.432359
600,0.489800,No Log,No Log,No Log
924,0.489800,0.392600,0.518471,0.488095
1200,0.396100,No Log,No Log,No Log
1386,0.396100,0.401789,0.532021,0.483225
1800,0.358700,No Log,No Log,No Log
1848,0.358700,0.401724,0.543318,0.485390
2310,0.358700,0.420338,0.520942,0.528139
2400,0.322800,No Log,No Log,No Log
2772,0.322800,0.426949,0.524082,0.514610



===== Epoch 1 =====
Train - loss=0.4086, F1=0.5254, Acc=0.4105
Val   - loss=0.4073, F1=0.5012, Acc=0.4324


===== Epoch 2 =====
Train - loss=0.3572, F1=0.5952, Acc=0.5042
Val   - loss=0.3926, F1=0.5185, Acc=0.4881


===== Epoch 3 =====
Train - loss=0.3273, F1=0.6375, Acc=0.5179
Val   - loss=0.4018, F1=0.5320, Acc=0.4832


===== Epoch 4 =====
Train - loss=0.3033, F1=0.6585, Acc=0.5305
Val   - loss=0.4017, F1=0.5433, Acc=0.4854


===== Epoch 5 =====
Train - loss=0.2874, F1=0.6823, Acc=0.5865
Val   - loss=0.4203, F1=0.5209, Acc=0.5281


===== Epoch 6 =====
Train - loss=0.2712, F1=0.7009, Acc=0.5854
Val   - loss=0.4269, F1=0.5241, Acc=0.5146


===== Epoch 7 =====
Train - loss=0.2679, F1=0.7069, Acc=0.5849
Val   - loss=0.4351, F1=0.5335, Acc=0.5184



[I 2025-11-25 17:22:56,309] Trial 16 finished with value: 0.5433176889496768 and parameters: {'learning_rate': 1.5705134127086373e-05, 'num_train_epochs': 7, 'per_device_train_batch_size': 16, 'weight_decay': 0.04703085746394833, 'warmup_ratio': 0.07204629975258536, 'dropout': 0.15385788772690773}. Best is trial 16 with value: 0.5433176889496768.


Final eval metrics: {'eval_loss': 0.4351390600204468, 'eval_f1_macro': 0.5335308965764392, 'eval_accuracy': 0.5183982683982684, 'eval_runtime': 2.4901, 'eval_samples_per_second': 742.152, 'eval_steps_per_second': 46.585, 'epoch': 7.0}


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.409260,0.480310,0.462121



===== Epoch 1 =====
Train - loss=0.4084, F1=0.4982, Acc=0.4475


[I 2025-11-25 17:24:13,173] Trial 17 pruned. 


Val   - loss=0.4093, F1=0.4803, Acc=0.4621

Pruning trial at epoch 1 with val F1=0.4803


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.417951,0.476381,0.453463



===== Epoch 1 =====
Train - loss=0.4259, F1=0.5006, Acc=0.4391
Val   - loss=0.4180, F1=0.4764, Acc=0.4535

Pruning trial at epoch 1 with val F1=0.4764


[I 2025-11-25 17:25:29,841] Trial 18 pruned. 
Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.415814,0.496439,0.410173
600,0.486700,No Log,No Log,No Log
924,0.486700,0.399352,0.505657,0.496753



===== Epoch 1 =====
Train - loss=0.4199, F1=0.5173, Acc=0.3817
Val   - loss=0.4158, F1=0.4964, Acc=0.4102


===== Epoch 2 =====
Train - loss=0.3682, F1=0.5726, Acc=0.5038


[I 2025-11-25 17:28:01,976] Trial 19 pruned. 


Val   - loss=0.3994, F1=0.5057, Acc=0.4968

Pruning trial at epoch 2 with val F1=0.5057
Best F1: 0.5433176889496768
Best params: {'learning_rate': 1.5705134127086373e-05, 'num_train_epochs': 7, 'per_device_train_batch_size': 16, 'weight_decay': 0.04703085746394833, 'warmup_ratio': 0.07204629975258536, 'dropout': 0.15385788772690773}


Best F1: 0.5433176889496768
Best params: {'learning_rate': 1.5705134127086373e-05, 'num_train_epochs': 7, 'per_device_train_batch_size': 16, 'weight_decay': 0.04703085746394833, 'warmup_ratio': 0.07204629975258536, 'dropout': 0.15385788772690773}

In [14]:
import gc
import torch
from transformers import TrainingArguments, Trainer
from transformers import DataCollatorWithPadding

# ---- Best hyperparameters from earlier good trial ----
best_params = {
    "learning_rate": 1.0319253484640193e-05,
    "num_train_epochs": 8,
    "per_device_train_batch_size": 16,
    "weight_decay": 0.06597712768869711,
    "warmup_ratio": 0.05106958730111153,
    "dropout": 0.15342334457930773,
}

# Optional: reproducibility
trial_seed = GLOBAL_SEED  # or any fixed int, e.g. 42
seed_everything(trial_seed)

# ---- Build model with fixed dropout ----
model = build_model(dropout=best_params["dropout"])

training_args = TrainingArguments(
    output_dir="./xlmr_single_run",  # new folder for this sanity run
    num_train_epochs=best_params["num_train_epochs"],
    learning_rate=best_params["learning_rate"],
    per_device_train_batch_size=best_params["per_device_train_batch_size"],
    per_device_eval_batch_size=best_params["per_device_train_batch_size"],
    save_strategy="no",
    logging_steps=600,          # adjust if you want more/less frequent logs
    report_to="none",
    weight_decay=best_params["weight_decay"],
    warmup_ratio=best_params["warmup_ratio"],
    metric_for_best_model="f1_macro",
    load_best_model_at_end=False,   # we’re just running straight through
    fp16=torch.cuda.is_available(),
    seed=trial_seed,
    data_seed=trial_seed,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=compute_metrics_multilabel,
    data_collator=DataCollatorWithPadding(tokenizer),
)

# Optional: attach the epoch-wise logging callback
epoch_cb = EpochMetricsCallback(
    trainer=trainer,
    train_dataset=train_dataset,
    val_dataset=val_dataset,
    trial=None,  # no Optuna here
)
trainer.add_callback(epoch_cb)

print("Trainer device:", trainer.args.device)

# ---- Run training + final eval ----
try:
    trainer.train()
    final_eval = trainer.evaluate()
    print("Final eval metrics:", final_eval)
finally:
    del trainer, model
    gc.collect()
    if torch.cuda.is_available():
        torch.cuda.empty_cache()


Some weights of XLMRobertaForSequenceClassification were not initialized from the model checkpoint at cardiffnlp/twitter-xlm-roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Trainer device: cuda:0


Step,Training Loss,Validation Loss,F1 Macro,Accuracy
462,No log,0.422920,0.442477,0.504870
600,0.496500,No Log,No Log,No Log
924,0.496500,0.405227,0.499382,0.489719
1200,0.411800,No Log,No Log,No Log
1386,0.411800,0.401951,0.512624,0.484307
1800,0.378300,No Log,No Log,No Log
1848,0.378300,0.408810,0.519961,0.500000
2310,0.378300,0.421438,0.517059,0.515693
2400,0.354800,No Log,No Log,No Log
2772,0.354800,0.423272,0.528307,0.507035



===== Epoch 1 =====
Train - loss=0.4185, F1=0.4835, Acc=0.4965
Val   - loss=0.4229, F1=0.4425, Acc=0.5049


===== Epoch 2 =====
Train - loss=0.3794, F1=0.5642, Acc=0.4921
Val   - loss=0.4052, F1=0.4994, Acc=0.4897


===== Epoch 3 =====
Train - loss=0.3514, F1=0.6021, Acc=0.5088
Val   - loss=0.4020, F1=0.5126, Acc=0.4843


===== Epoch 4 =====
Train - loss=0.3304, F1=0.6284, Acc=0.5370
Val   - loss=0.4088, F1=0.5200, Acc=0.5000


===== Epoch 5 =====
Train - loss=0.3161, F1=0.6540, Acc=0.5604
Val   - loss=0.4214, F1=0.5171, Acc=0.5157


===== Epoch 6 =====
Train - loss=0.3009, F1=0.6675, Acc=0.5547
Val   - loss=0.4233, F1=0.5283, Acc=0.5070


===== Epoch 7 =====
Train - loss=0.2930, F1=0.6784, Acc=0.5696
Val   - loss=0.4276, F1=0.5302, Acc=0.5211


===== Epoch 8 =====
Train - loss=0.2912, F1=0.6812, Acc=0.5750
Val   - loss=0.4348, F1=0.5298, Acc=0.5222



Final eval metrics: {'eval_loss': 0.43482330441474915, 'eval_f1_macro': 0.529784192477245, 'eval_accuracy': 0.5221861471861472, 'eval_runtime': 2.5029, 'eval_samples_per_second': 738.356, 'eval_steps_per_second': 46.347, 'epoch': 8.0}
